### PR TITLE
Benchmark for ST_Contains

### PR DESCRIPTION
**Pull request by mbasmanova**
It turns out that 60% of execution time for ST_Contains is spent deserializing geometries from slices. This PR adds a benchmark to show that and optimizes ST_Contains to avoid costly deserialization when possible.

The benchmark runs ST_Contains for a large polygon and 3 points:
* inside the polygon, 
* outside of the polygon, but inside the polygon's bounding box,
* outside of the polygon's bounding box. 

The benchmark runs ST_Contains using both slices and geometries. Here are the results:

```
Benchmark                                                            Mode  Cnt       Score       Error  Units
BenchmarkSTContains.deserialize                                      avgt   20  131919.856 ±  4921.112  ns/op
BenchmarkSTContains.deserializeSimpleGeometry                        avgt   20     486.274 ±    13.489  ns/op
BenchmarkSTContains.stContainsInnerPoint                             avgt   20  227284.106 ± 18774.481  ns/op
BenchmarkSTContains.stContainsInnerPointDeserialized                 avgt   20   66938.920 ±  1570.299  ns/op
BenchmarkSTContains.stContainsInnerPointSimpleGeometry               avgt   20     813.359 ±    34.580  ns/op
BenchmarkSTContains.stContainsOuterPointInEnvelope                   avgt   20  218173.265 ±  9345.754  ns/op
BenchmarkSTContains.stContainsOuterPointInEnvelopeDeserialized       avgt   20   69004.912 ±  2625.112  ns/op
BenchmarkSTContains.stContainsOuterPointInEnvelopeSimpleGeometry     avgt   20     837.565 ±    58.022  ns/op
BenchmarkSTContains.stContainsOuterPointNotInEnvelope                avgt   20  134204.161 ±  5208.045  ns/op
BenchmarkSTContains.stContainsOuterPointNotInEnvelopeDeserialized    avgt   20      30.272 ±     1.232  ns/op
BenchmarkSTContains.stContainsOuterPointNotInEnvelopeSimpleGeometry  avgt   20     643.869 ±    15.003  ns/op
```

For points within the polygon's bounding box, ST_Contains on geometries is almost 3 times faster than on slices. For points outside of the polygon's bounding, the difference is even larger.

ST_Contains runs in two phases: (1) check if bounding box of the left geometry contains bounding box of the right geometry; if not, return false; (2) check if the left geometry contains the right geometry. This is what makes ST_Contains for points outside of polygon's bounding box run faster.

Phase 1 doesn't require full geometries, only their bounding boxes. Hence, it is sufficient to deserialize just that part (read 4 doubles). Making this change greatly speeds up ST_Contains for cases when computation completes in phase 1 and doesn't proceed to phase 2 (e.g. point is outside polygon's bounding box). Benchmark results for the optimized version of the
ST_Contains show 6x improvement for simple geometry and 1000x improvement for
complex geometry.

```
Benchmark                                                            Mode  Cnt       Score       Error  Units
BenchmarkSTContains.deserialize                                      avgt   20  138402.816 ± 11643.730  ns/op
BenchmarkSTContains.deserializeEnvelope                              avgt   20      59.143 ±     2.541  ns/op
BenchmarkSTContains.deserializeEnvelopeSimpleGeometry                avgt   20      61.340 ±     3.933  ns/op
BenchmarkSTContains.deserializeSimpleGeometry                        avgt   20     499.509 ±    22.847  ns/op
BenchmarkSTContains.stContainsInnerPoint                             avgt   20  231484.522 ± 17018.650  ns/op
BenchmarkSTContains.stContainsInnerPointDeserialized                 avgt   20   67700.512 ±  2490.978  ns/op
BenchmarkSTContains.stContainsInnerPointSimpleGeometry               avgt   20     920.532 ±    21.465  ns/op
BenchmarkSTContains.stContainsOuterPointInEnvelope                   avgt   20  206741.486 ±  3162.731  ns/op
BenchmarkSTContains.stContainsOuterPointInEnvelopeDeserialized       avgt   20   66869.945 ±  1507.815  ns/op
BenchmarkSTContains.stContainsOuterPointInEnvelopeSimpleGeometry     avgt   20     919.686 ±    28.572  ns/op
BenchmarkSTContains.stContainsOuterPointNotInEnvelope                avgt   20     109.417 ±     2.514  ns/op
BenchmarkSTContains.stContainsOuterPointNotInEnvelopeDeserialized    avgt   20      29.153 ±     0.734  ns/op
BenchmarkSTContains.stContainsOuterPointNotInEnvelopeSimpleGeometry  avgt   20     109.529 ±     2.475  ns/op
```

### CODE REVIEW

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
+package com.facebook.presto.plugin.geospatial;
```
Comment (kokosing) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java]: copyright

Comment (sopel39) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java]: Could you add benchmarks result in commit message?

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Thread)
+@Fork(2)
+@Warmup(iterations = 5, time = 500, timeUnit = TimeUnit.MILLISECONDS)
```
Comment (sopel39) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java]: Average error seems rather high, e.g:
```
BenchmarkSTContains.stContainsInnerPointDeserialized               avgt   10   0.093 ±  0.045  ms/op
...
BenchmarkSTContains.stContainsOuterPointInEnvelopeDeserialized     avgt   10   0.094 ±  0.047  ms/op
```
consider more iterations.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        private OGCGeometry ogcGeometry;
+        private OGCPoint innerOgcPoint;
+        private OGCPoint outerOgcPointInEnvelope;
+        private OGCPoint outerOgcPointNotInEnvelope;
+
+        @Setup
+        public void setup()
+                throws IOException
+        {
+            Path filePath = Paths.get(this.getClass().getClassLoader().getResource("large_polygon.txt").getPath());
```
Comment (sopel39) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java]: `large_polygon.txt` looks complex. Do you think it would be worth benchmarking with smaller/simpler polygons?

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        }
+    }
+
+    public static void main(String[] args)
+            throws IOException, RunnerException
+    {
+        // assure the benchmarks are valid before running
+        BenchmarkData data = new BenchmarkData();
+        data.setup();
+        new BenchmarkSTContains().stContainsInnerPoint(data);
```
Comment (sopel39) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java]: throw if not contains?

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 5]
```diff
@@ -73,6 +74,42 @@ public static GeometryTypeName valueOf(String type)
...
+                double ymax = buffer.getDouble();
+                envelope = new Envelope(xmin, ymin, xmax, ymax);
+            }
+            if (overallEnvelope == null) {
+                overallEnvelope = envelope;
+            } else {
+                overallEnvelope.merge(envelope);
+            }
+        }
+
```
Comment (sopel39) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: ensure `overallEnvelope != null`?

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 6]
```diff
@@ -73,6 +74,42 @@ public static GeometryTypeName valueOf(String type)
         }
     }
 
+    public static Envelope deserializeEnvelope(Slice shape)
```
Comment (sopel39) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: Would it be possible that `Slice shape` contains raw data for envelope. It seems that `Envelope#merge(com.esri.core.geometry.Envelope)` might have quadratic complexity.

Comment (sopel39) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: Could you add benchmark of this new method?

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java (hunk 2 above)
Comment (mbasmanova) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTContains.java]: @sopel39 , thanks for the suggestion. With 10 iterations the error is very small:

```
Benchmark                                                          Mode  Cnt   Score    Error  Units
BenchmarkSTContains.deserialize                                    avgt   20   0.131 ±  0.004  ms/op
BenchmarkSTContains.deserializeEnvelope                            avgt   20  ≈ 10⁻⁴           ms/op
BenchmarkSTContains.stContainsInnerPoint                           avgt   20   0.205 ±  0.008  ms/op
BenchmarkSTContains.stContainsInnerPointDeserialized               avgt   20   0.064 ±  0.001  ms/op
BenchmarkSTContains.stContainsOuterPointInEnvelope                 avgt   20   0.210 ±  0.007  ms/op
BenchmarkSTContains.stContainsOuterPointInEnvelopeDeserialized     avgt   20   0.066 ±  0.002  ms/op
BenchmarkSTContains.stContainsOuterPointNotInEnvelope              avgt   20  ≈ 10⁻⁴           ms/op
BenchmarkSTContains.stContainsOuterPointNotInEnvelopeDeserialized  avgt   20  ≈ 10⁻⁵           ms/op
```

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java (hunk 5 above)
Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: overallEnvelope can be null if geometry is empty (e.g. 'POINT EMPTY', 'POLYGON EMPTY', etc.); I'll add @Nullable to the signature

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java (hunk 6 above)
Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: @sopel39 , at the moment we have little control over serde of the geometry type, because we use serde from the ESRI library. I do keep running into cases where custom serde fully controlled in Presto would allow for more optimizations, but for now I want to keep it as is.  `Slice shape` contains raw data for the envelope for all shape types except for geometry collection which is rarely used.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 7]
```diff
@@ -73,6 +76,57 @@ public static GeometryTypeName valueOf(String type)
...
+        Envelope overallEnvelope = null;
+        while (input.available() > 0) {
+            int length = input.readInt();
+            ByteBuffer buffer = input.readSlice(length).toByteBuffer().slice().order(LITTLE_ENDIAN);
+            int type = buffer.getInt();
+            Envelope envelope = null;
+            if (type == 1) {    // point
+                double x = buffer.getDouble();
+                double y = buffer.getDouble();
+                if (!Double.isNaN(x) && !Double.isNaN(translateFromAVNaN(x))) {
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: What if `x` or `xmin` below is NaN? Should we checkArgument/verify for these?

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 8]
```diff
@@ -73,6 +76,57 @@ public static GeometryTypeName valueOf(String type)
...
+            return null;
+        }
+        BasicSliceInput input = shape.getInput();
+
+        input.readInt(); // spatial reference ID
+
+        Envelope overallEnvelope = null;
+        while (input.available() > 0) {
+            int length = input.readInt();
+            ByteBuffer buffer = input.readSlice(length).toByteBuffer().slice().order(LITTLE_ENDIAN);
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: Why do we need the `slice()` call here?

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 9]
```diff
@@ -73,6 +76,57 @@ public static GeometryTypeName valueOf(String type)
...
+            int type = buffer.getInt();
+            Envelope envelope = null;
+            if (type == 1) {    // point
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: we can define a constant like `private static final int POINT_TYPE = 1;`

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 10]
```diff
@@ -73,6 +76,57 @@ public static GeometryTypeName valueOf(String type)
...
+                double x = buffer.getDouble();
+                double y = buffer.getDouble();
+                if (!Double.isNaN(x) && !Double.isNaN(translateFromAVNaN(x))) {
+                    envelope = new Envelope(x, y, x, y);
+                }
+            }
+            else {
+                double xmin = buffer.getDouble();
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: camelcase like `xMin` etc.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 11]
```diff
@@ -73,6 +76,57 @@ public static GeometryTypeName valueOf(String type)
         }
     }
 
+    // Copy of com.esri.core.geometry.Interop.translateFromAVNaN
+    private static double translateFromAVNaN(double n)
```
Comment (sopel39) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: why is that needed?

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 12]
```diff
@@ -73,6 +76,57 @@ public static GeometryTypeName valueOf(String type)
...
+                if (!Double.isNaN(x) && !Double.isNaN(translateFromAVNaN(x))) {
+                    envelope = new Envelope(x, y, x, y);
+                }
+            }
+            else {
+                double xmin = buffer.getDouble();
+                double ymin = buffer.getDouble();
+                double xmax = buffer.getDouble();
+                double ymax = buffer.getDouble();
+                if (!Double.isNaN(xmin) && !Double.isNaN(translateFromAVNaN(xmin))) {
```
Comment (sopel39) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: ditto what, if `ymin, ymax` is `NaN`

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 13]
```diff
@@ -73,6 +76,57 @@ public static GeometryTypeName valueOf(String type)
...
+                    envelope = new Envelope(xmin, ymin, xmax, ymax);
+                }
+            }
+            if (envelope != null) {
```
Comment (sopel39) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: I understand `envelope` might be `null` if some coordinate is `NaN`. In what case can this happen? Don't we loose some valuable info this way?

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoQueries.java [hunk 14]
```diff
@@ -331,6 +335,20 @@ public void testSTSymmetricDifference()
         assertFunction("ST_AsText(ST_SymDifference(ST_GeometryFromText('MULTIPOLYGON (((0 0 , 0 2, 2 2, 2 0)), ((2 2, 2 4, 4 4, 4 2)))'), ST_GeometryFromText('POLYGON ((0 0, 0 3, 3 3, 3 0))')))", VARCHAR, "MULTIPOLYGON (((2 0, 3 0, 3 2, 2 2, 2 0)), ((0 2, 2 2, 2 3, 0 3, 0 2)), ((3 2, 4 2, 4 4, 2 4, 2 3, 3 3, 3 2)))");
     }
 
+    @Test
+    public void testDeserializeEnvelope()
+    {
+        assertEquals(deserializeEnvelope(stGeometryFromText(Slices.utf8Slice("POINT (1 2)"))), new Envelope(1, 2, 1, 2));
```
Comment (sopel39) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoQueries.java]: can you add cases for `NaN` envelope coordinates?

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java (hunk 8 above)
Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: We don't need it here. I copy-pasted that part from deserialize(). I'll remove it.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java (hunk 11 above)
Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: deserializeEnvelope needs to recognize custom NAN values generated by ESRI's serialization of empty geometries.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java (hunk 12 above)
Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: added Verify.verify calls

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java (hunk 13 above)
Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: NaN values are used when serializing empty geometries, e.g. 'POINT EMPTY', 'POLYGON EMPTY'.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoQueries.java (hunk 14 above)
Comment (mbasmanova) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoQueries.java]: I have these already:

```
assertEquals(deserializeEnvelope(stGeometryFromText(Slices.utf8Slice("POLYGON EMPTY"))), null);
assertEquals(deserializeEnvelope(stGeometryFromText(Slices.utf8Slice("POINT EMPTY"))), null);
```

**Comment by mbasmanova** [comment]
@sopel,

>> It seems (from results) that optimized stContains is slightly slower in general case, but helps greatly in extreme cases.

For shapes on the left which lie within the bounding box of the shape on the right optimized ST_Contains is just under 2% slower. For shapes that don't lie within the bounding box the new version can be thousands of times faster. I think of both cases as common scenarios. Also, I'm seeing many queries where the latter case happens a lot more often than the former. Consider a spatial join `SELECT ... FROM a, b WHERE ST_Contains(a.x, b.y)` where ST_Contains is applied to all combinations of records from 'a' and 'b'. In this query only tiny subset of invocations will require checking beyond bounding boxes. A range query filtering a dataset using fixed polygon would be similar.

I looked at a concrete example where table 'a' had 800K rows and table 'b' had 34. ST_Contains is invoked 27M times. In 26M of these, bounding boxes don't match. In 500K cases bounding boxes do match. The old implementation would spend about 1000 hours (non-distributed), while the new one will spend only 28 hours.

```
  New: 204ms x 500K + 0.1ms x 26M ~= 104K seconds 
  Old: 200ms x 500K + 134ms x 26M ~= 1K hours
```

**Comment by mbasmanova** [comment]
@nezihyigitbasi , @sopel39 , thanks for reviews. I believe I addressed all comments.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 15]
```diff
@@ -13,15 +13,19 @@
  */
 package com.facebook.presto.geospatial;
 
+import com.esri.core.geometry.Envelope;
 import com.esri.core.geometry.Geometry;
 import com.esri.core.geometry.GeometryCursor;
 import com.esri.core.geometry.SpatialReference;
 import com.esri.core.geometry.ogc.OGCConcreteGeometryCollection;
 import com.esri.core.geometry.ogc.OGCGeometry;
+import com.google.common.base.Verify;
```
Comment (sopel39) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: I think we used to use `checkState` from `Preconditions`. `Verify` is tagged as `@Beta` class.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java [hunk 16]
```diff
@@ -73,6 +79,71 @@ public static GeometryTypeName valueOf(String type)
...
+                    envelope = new Envelope(x, y, x, y);
+                }
+            }
+            else {
+                double xMin = buffer.getDouble();
+                double yMin = buffer.getDouble();
+                double xMax = buffer.getDouble();
+                double yMax = buffer.getDouble();
+                if (!isEsriNaN(xMin)) {
+                    Verify.verify(!isEsriNaN(xMax));
```
Comment (sopel39) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: I wonder if it might have some perf impact...

Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: @nezihyigitbasi , @martint , do you know whether we should use Verify.verify or Preconditions.checkState ?

Comment (nezihyigitbasi) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: If you are asking in general when to use which one, please see [this](https://github.com/google/guava/wiki/ConditionalFailuresExplained). There is also some good info in the `Verify` [javadoc](https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/base/Verify.html).

Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java]: @nezihyigitbasi , thank you for the pointers.
